(DOCUMENT-773) Correct "Calling other functions" for custom Ruby functions

nrwahl2 opened the pull request:
Before this update, the "Calling other functions" section of
functions_ruby_implementation.md incorrectly described a 3.x
style function call (`call_function(func_name, args, &block)`).
This patch corrects the documentation to reflect the 4.x style:
`call_function(func_name, *args, &block)`.

puppetcla commented:
CLA signed by all contributors.

hlindberg commented on source/puppet/4.10/functions_ruby_implementation.md:
```diff
@@ -77,28 +77,28 @@ end
 * The first argument must be the name of the function to call, as a string.
-* The second argument must be an array containing any arguments to the function.
-* The third argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
+* The next arguments can be any data type that the called function accepts. They will be passed through directly as arguments to the called function.
```
Not sure why the phrase "passed through directly" is saying - it may not be true as there may be type casting involved if a parameter type is of `Init` type. Suggest taking out those words.
(Also, it opens up the question "what is the difference between direct and indirect passing of arguments - is there even such a thing?").

hlindberg commented on source/puppet/4.10/functions_ruby_implementation.md:
```diff
@@ -77,28 +77,28 @@ end
-* The second argument must be an array containing any arguments to the function.
-* The third argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
+* The next arguments can be any data type that the called function accepts. They will be passed through directly as arguments to the called function.
+* The final argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
```
Maybe say "last argument" since "being final" may associate to the meaning "not overridable"?

hlindberg commented on source/puppet/4.9/functions_ruby_implementation.md:
```diff
@@ -77,28 +77,28 @@ end
-* The second argument must be an array containing any arguments to the function.
-* The third argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
+* The next arguments can be any data type that the called function accepts. They will be passed through directly as arguments to the called function.
+* The final argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
```
Same comments here as for the above...

hlindberg commented on source/puppet/5.0/functions_ruby_implementation.md:
```diff
@@ -77,28 +77,28 @@ end
-* The second argument must be an array containing any arguments to the function.
-* The third argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
+* The next arguments can be any data type that the called function accepts. They will be passed through directly as arguments to the called function.
+* The final argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
```
And here...

hlindberg commented on source/puppet/5.1/functions_ruby_implementation.md:
```diff
@@ -77,28 +77,28 @@ end
-* The second argument must be an array containing any arguments to the function.
-* The third argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
+* The next arguments can be any data type that the called function accepts. They will be passed through directly as arguments to the called function.
+* The final argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
```
again...

hlindberg commented on source/puppet/5.2/functions_ruby_implementation.md:
```diff
@@ -77,28 +77,28 @@ end
-* The second argument must be an array containing any arguments to the function.
-* The third argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
+* The next arguments can be any data type that the called function accepts. They will be passed through directly as arguments to the called function.
+* The final argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
```
same...

hlindberg commented on source/puppet/5.3/functions_ruby_implementation.md:
```diff
@@ -77,28 +77,28 @@ end
-* The second argument must be an array containing any arguments to the function.
-* The third argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
+* The next arguments can be any data type that the called function accepts. They will be passed through directly as arguments to the called function.
+* The final argument can be a Ruby Proc, or a Puppet [lambda][lambdas] previously captured as a Proc (see above). You can also provide a block of Ruby code using the normal block syntax.
```
same...

nrwahl2 commented:
This confusion stemmed from my lack of understanding of Ruby Proc/block syntax. I'm going to hold off on touching anything I don't understand for a little while after this, since I don't have a lot of cycles to look at GitHub at the moment. I'm submitting an update now that addresses your comments.

hlindberg commented:
@nrwahl2 Looks good now... merging.
